### PR TITLE
AWS RDS Monitoring via Icinga

### PR DESCRIPTION
**Pull request by suthagarht**
- We have a requirement to monitor AWS RDS instances via the current
monitoring system (Icinga). We have made few changes to accomplish this
objective.

- We have created check scripts as plugins. These have been placed
inside 'modules/monitoring/files/usr/lib/nagios/plugins/'.

- We have created config files that will translate into the actual
command and parameters. This has been placed inside 'modules/monitoring/files/etc/nagios3/conf.d/'.

- We have created a new class called 'monitoring::checks::rds' to create
the necessary files in the monitoring instance. This resides inside 'modules/monitoring/manifests/checks/rds.pp'.

- We have enabled the class by calling it at 'modules/monitoring/manifests/checks.pp'.

- We have specified the server names in 'hieradata_aws/common.yaml'.

https://trello.com/c/C7SY8edZ/1123-monitor-rds-and-elasticache-in-aws-and-alert-icinga

Pair: @suthagarht @szd55gds

### CODE REVIEW

File: modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_cpu [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+#################################################
+# This script will cpture and report on AWS RDS #
+# CPUUtilization                                #
+#################################################
+
+#######################################################################################################
+# Authors : GOV.UK Reliability Engineering                                                            #
+# Date : 06/06/2018                                                                                   #
+# Version : 01                                                                                        #
```
[the remark this answers is not in the document]
Comment (suthagarht) [reply modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_cpu]: done

File: modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_cpu [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+# Source : govuk-puppet/tree/master/modules/monitoring/files/etc/nagios3/conf.d/check_aws_rds_cpu.cfg #
```
[the remark this answers is not in the document]
Comment (suthagarht) [reply modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_cpu]: agreed :)

File: modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_cpu [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+# 3) If 1 and 2 are correct, check whether the average    #
+#    CPUUtilaization is in the WARNING or CRITICAL state. #
+#    Else, mark the status as OK.                         #
+###########################################################
+
+###############################################################################
+# How to use and test this script?                                            #
+# --------------------------------                                            #
+#                                                                             #
+# This script takes input from the following file when it is used by icinga.  #
```
[the remark this answers is not in the document]
Comment (suthagarht) [reply modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_cpu]: Good question. The ID change doesn't actually have an impact. According to puppet, icinga is told where to look for when host config is required.

File: modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_cpu [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+#                                                                             # 
+# Test                                                                        #
+# ----                                                                        #
+# You can test this script by following the steps below.                      #
+#                                                                             #
+# 1) Connect to the 'blue-monitoring' instance via ssh.                       #
+# 2) Become root. (sudo -i)                                                   #
+# 3) Change directory. (cd /usr/lib/nagios/plugins)                           #
+# 4) Execute the script with manual input.                                    #
+#    - r = eu-west-1  (Example AWS Staging region)                            #
```
[the remark this answers is not in the document]
Comment (suthagarht) [reply modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_cpu]: done

File: modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_memory [hunk 5]
```diff
@@ -0,0 +1,199 @@
+#!/usr/bin/env python
+
+#################################################
+# This script will cpture and report on AWS RDS #
```
[the remark this answers is not in the document]
Comment (suthagarht) [reply modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_memory]: sorry

File: modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_memory [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+parser.add_argument("-r","--region", type=str, choices=['us-east-1','eu-west-1'], help="AWS region to check")
+parser.add_argument("-w","--warning", type=float, help="Value(GB) at which to raise a warning alert (5.0)", default=5.0)
+parser.add_argument("-c","--critical", type=float, help="Value(GB) at which to raise a critical alert (2.0)", default=2.0)
+parser.add_argument("-i","--instanceid", type=str, help="Instanceid", default=0)
+args = parser.parse_args()
+
+############################################
+# Check whether a region has been provided.#
+############################################
+if args.region is None:
```
[the remark this answers is not in the document]
Comment (suthagarht) [reply modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_memory]: If an invalid region is passed ..

error: argument -r/--region: invalid choice: 'eu-west-45' (choose from 'us-east-1', 'eu-west-1')

File: modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_memory [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+##########################################
+cw_conn = boto.ec2.cloudwatch.connect_to_region(args.region)
+
+###########################################################
+# Define start and end times      			  #
+# Here we have give 5 minutes because that is the window  #
+# we are using to calculate the average.		  #
+# Note: The time should be in UTC format.		  #
+###########################################################
+start = datetime.datetime.now() - datetime.timedelta(seconds=300)
```
[the remark this answers is not in the document]
Comment (suthagarht) [reply modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_memory]: sorry, I fail to see a very strong need to change this. Please do contradict if necessary.

File: modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_memory [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+end = datetime.datetime.now()
+
+
+#########################################################################
+# Use the cloutwatch connection and get metrics.           		#
```
[the remark this answers is not in the document]
Comment (suthagarht) [reply modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_memory]: done :)

File: modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_memory [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+# 'end'  - This will pass the end time.					#
+# 'FreeableMemory' - This is the matric that we are asking for from	#
+#		     CloudWatch.					#
+# 'AWS/RDS' - This is the CloudWatch space/area where RDS stats reside. #
+# 'Average' - We are defining that we want the average for the matric.	#
+# 'DBInstanceIdentifier' - The unique identifier which we use to query. #
+#########################################################################
+
+data = cw_conn.get_metric_statistics(
+   300,
```
[the remark this answers is not in the document]
Comment (suthagarht) [reply modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_memory]: done

File: modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_memory [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+#########################################
+# Disect the result			#
+# -----------------			#
+# The result that we get is a 'list' of	#
+# 'dictonary' elements. Hence, we will	#
+# iterate over the list and get the 	#
+# value for the key called "Average"	#
+#########################################
+for item in data:
+ memory_bytes = item["Average"]
```
[the remark this answers is not in the document]
Comment (suthagarht) [reply modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_memory]: added soft tab :)

File: modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_memory [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+
+# The average value is returned in bytes. Here we are converting it to giga bytes.
+memory = memory_bytes / (1024 * 1024 * 1024)
+print(memory)
```
[the remark this answers is not in the document]
Comment (suthagarht) [reply modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_memory]: yup, debugging forgotten :)

File: modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_memory [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+
+#########################################################
+# Print the output					#
+# This is the final step. We do the following here.	#
+# 1) Change the average vale to 'float' type.		#
+# 2) Convert the defined value to 'float' type.		#
+# 3) Compare the check vs actual.			#
+#########################################################
+if float(memory) < float(args.critical):
```
[the remark this answers is not in the document]
Comment (suthagarht) [reply modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_memory]: yup, corrected

File: modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_storage [hunk 13]
```diff
@@ -0,0 +1,198 @@
+#!/usr/bin/env python
+
+#################################################
+# This script will cpture and report on AWS RDS #
+# FreeStorageSpace                              #
+#################################################
+
+###########################################################################################################
+# Authors : GOV.UK Reliability Engineering                                                            	  #
```
[the remark this answers is not in the document]
Comment (suthagarht) [reply modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_storage]: done

File: modules/monitoring/manifests/checks/rds.pp [hunk 14]
```diff
@@ -0,0 +1,48 @@
+# == Class: monitoring::checks::rds
+#
+# Nagios alerts for AWS RDS.
+#
+# === Parameters
+#
```
[the remark this answers is not in the document]
Comment (suthagarht) [reply modules/monitoring/manifests/checks/rds.pp]: correct, done

File: modules/monitoring/manifests/checks/rds.pp [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    icinga::plugin { 'check_aws_rds_storage':
+        source  => 'puppet:///modules/monitoring/usr/lib/nagios/plugins/check_aws_rds_storage',
+        require => Package['boto'],
+    }
+
+    icinga::check_config { 'check_aws_rds_storage':
+        source  => 'puppet:///modules/monitoring/etc/nagios3/conf.d/check_aws_rds_storage.cfg',
+        require => File['/usr/lib/nagios/plugins/check_aws_rds_cpu'],
```
[the remark this answers is not in the document]
Comment (suthagarht) [reply modules/monitoring/manifests/checks/rds.pp]: excellent, not sure how I missed or how you spotted 👏

File: modules/monitoring/manifests/checks/rds.pp [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    if $enabled {
+    monitoring::checks::rds_config { $servers: }
```
[the remark this answers is not in the document]
Comment (suthagarht) [reply modules/monitoring/manifests/checks/rds.pp]: done

File: modules/monitoring/manifests/checks/rds_config.pp [hunk 17]
```diff
@@ -0,0 +1,55 @@
+# == Define: monitoring::checks::rds_config
+#
+# Nagios alert config for AWS RDS.
+#
+# === Parameters
+#
+# [*region*]
+#   Which AWS region should be checked ['us-east-1','eu-west-1']
```
[the remark this answers is not in the document]
Comment (suthagarht) [reply modules/monitoring/manifests/checks/rds_config.pp]: corrected

File: modules/monitoring/manifests/checks/rds_config.pp [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+#
+define monitoring::checks::rds_config (
+  $region = undef,
+  $cpu_warning = 80,
+  $cpu_critical = 90,
+  $memory_warning = 5,
+  $memory_critical = 2,
+  $storage_warning = 20,
+  $storage_critical = 10,
+  ){
```
[the remark this answers is not in the document]
Comment (suthagarht) [reply modules/monitoring/manifests/checks/rds_config.pp]: Sorry, the others are and the check fails without the indentation.

File: modules/monitoring/manifests/checks/rds.pp [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+#   List of RDS instances.
+#
+class monitoring::checks::rds (
+      $enabled = true,
+      $servers = [],
+    ) {
+
+    icinga::plugin { 'check_aws_rds_cpu':
+        source  => 'puppet:///modules/monitoring/usr/lib/nagios/plugins/check_aws_rds_cpu',
+        require => Package['boto'],
```
[the remark this answers is not in the document]
Comment (suthagarht) [reply modules/monitoring/manifests/checks/rds.pp]: I do agree

File: modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_memory (hunk 6 above)
[the remark this answers is not in the document]
Comment (deanwilson) [reply modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_memory]: You're right 👍

File: modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_memory (hunk 7 above)
[the remark this answers is not in the document]
Comment (deanwilson) [reply modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_memory]: This is fine. I'm being overly picky

File: modules/monitoring/manifests/checks/rds.pp (hunk 14 above)
[the remark this answers is not in the document]
Comment (deanwilson) [reply modules/monitoring/manifests/checks/rds.pp]: 👍

File: modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_cpu (hunk 3 above)
[the remark this answers is not in the document]
Comment (suthagarht) [reply modules/monitoring/files/usr/lib/nagios/plugins/check_aws_rds_cpu]: done

File: modules/monitoring/manifests/checks/rds_config.pp (hunk 18 above)
[the remark this answers is not in the document]
Comment (suthagarht) [reply modules/monitoring/manifests/checks/rds_config.pp]: done

**Pull request merged**